platoon/controller.py

```diff
@@ -0,0 +1,267 @@
+import os
+
+import numpy
+import posix_ipc
+import zmq
+from mpi4py import MPI
+
+from mmap import mmap
+
+
+class Controller(object):
+    """
+    Abstract multi-process controller
+
+    This class provides the necessary features to dispatch data mini-batches
+    to workers and handle control requests. Using this class should be done
+    by having another class inherit from it and override the method
+    `handle_control()`.
+
+    .. warning::
+
+        Due to the underlying implementation it is a bad idea to
+        attempt to do both in the same process, even on different
+        threads.  This will suffer from interlock problems and may
+        negate any speedup you could get from using multiple Workers.
+
+        Because of this issue, the class may be split in the future.
+
+    Parameters
+    ----------
+    port : int
+        The port number to communicate over
+    control_port : int
+        The control port number.
+    device_list : list of strings
+        Contains device names in clique order (prefer ring topology)
+    hwm : int
+        High water mark (see pyzmq docs).
+
+    """
+
+    def __init__(self, control_port, device_list, global_size, global_rank,
+                 port=None, hwm=10, global_comm_id=""):
+
+        self._should_stop = False
+        self._worker_list = set()
+        self._need_init = True
+
+        self._device_list = device_list
+        # Controllers' MPI comm info
+        self._global_rank = global_rank
+        self._global_size = global_size
+        self._global_comm_id = "platoon-global_comm_" + global_comm_id
+        self._global_comm = None
+        # These 3 variables below will be used to take advantage of multi-node
```

Comment:
where do you see that nccl support more then 1 node?

On Wed, Jul 20, 2016 at 2:44 PM, Christos Tsirigotis <
notifications@github.com> wrote:

> In platoon/controller.py
> <https://github.com/mila-udem/platoon/pull/66#discussion_r71583281>:
>
> > +    """
> > +
> > +    def __init__(self, control_port, device_list, global_size, global_rank,
> > +                 port=None, hwm=10, global_comm_id=""):
> > +
> > +        self._should_stop = False
> > +        self._worker_list = set()
> > +        self._need_init = True
> > +
> > +        self._device_list = device_list
> > +        # Controllers' MPI comm info
> > +        self._global_rank = global_rank
> > +        self._global_size = global_size
> > +        self._global_comm_id = "platoon-global_comm_" + global_comm_id
> > +        self._global_comm = None
> > +        # These 3 variables below will be used to take advantage of multi-node
>
> I will do. I have something in mind about taking advantage of nccl's
> 2-node support. But this is an optimization so I will do later
> … <#m_-2417918615462058764_>
> On Jul 20, 2016 19:44, "abergeron" ***@***.***> wrote: In
> platoon/controller.py <#66 (comment)
> <https://github.com/mila-udem/platoon/pull/66#discussion_r71561801>>: > +
> """ > + > + def __init__(self, control_port, device_list, global_size,
> global_rank, > + port=None, hwm=10, global_comm_id=""): > + > +
> self._should_stop = False > + self._worker_list = set() > + self._need_init
> = True > + > + self._device_list = device_list > + # Controllers' MPI comm
> info > + self._global_rank = global_rank > + self._global_size =
> global_size > + self._global_comm_id = "platoon-global_comm_" +
> global_comm_id > + self._global_comm = None > + # These 3 variables below
> will be used to take advantage of multi-node If they are not useful right
> now, then remove them, We have no idea how it will work later. — You are
> receiving this because you authored the thread. Reply to this email
> directly, view it on GitHub <
> https://github.com/mila-udem/platoon/pull/66/files/c46c24a835a8cb89f1f152d3d80b34e92d3703d4#r71561801>,
> or mute the thread <
> https://github.com/notifications/unsubscribe-auth/AD5VAeuEUUJge7Vga4p3Nc7MnRMvOvtsks5qXlCAgaJpZM4JOXc8>
> .
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/mila-udem/platoon/pull/66/files/c46c24a835a8cb89f1f152d3d80b34e92d3703d4#r71583281>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AALC-_zaMJctr95S8tcfmsdNFKYSgYjNks5qXmx6gaJpZM4JOXc8>
> .
>